protocols.py

```diff
@@ -1,67 +1,81 @@
-    try:
-        efi_defs = open(efi_def_path, "r").readlines()
-    except:
-        log_alert(f"Could not open EFI type definition file at '{efi_def_path}'. Your version of Binary Ninja may be out of date. Please update to version 3.5.4331 or higher.")
+    user_mapping_path = os.path.join(user_directory(), "types", "efi-protocol.json")
+
+    if not os.path.exists(bundled_efi_path):
+        log_alert(f"Could not open EFI type definition file at '{bundled_efi_path}'. Your version of Binary Ninja may be out of date. Please update to version 3.5.4331 or higher.")
         return False
 
-    protocols = {}
-
-    # Parse the GUID to protocol structure mappings out of the type definition source
-    guids = []
-    for line in efi_defs:
-        if line.startswith("///@protocol"):
-            guid = line.split("///@protocol")[1].replace("{", "").replace("}", "").strip().split(",")
-            guid = [int(x, 16) for x in guid]
-            guid = struct.pack("<IHHBBBBBBBB", *guid)
-            guids.append((guid, None))
-        elif line.startswith("///@binding"):
-            guid_name = line.split(" ")[1]
-            guid = line.split(" ")[2].replace("{", "").replace("}", "").strip().split(",")
-            guid = [int(x, 16) for x in guid]
-            guid = struct.pack("<IHHBBBBBBBB", *guid)
-            guids.append((guid, guid_name))
-        elif line.startswith("struct"):
-            name = line.split(" ")[1].strip()
-            for guid_info in guids:
-                guid, guid_name = guid_info
-                if guid_name is None:
-                    protocols[guid] = (name, f"{name}_GUID")
-                else:
-                    protocols[guid] = (name, guid_name)
-        else:
-            guids = []
+    with open(bundled_mapping_path, "r") as f:
+        bundled_mapping = json.load(f)
+        for mapping in bundled_mapping:
+            guid = struct.pack("<IHH8B", *mapping["guid"])
+            protocol_type = mapping["protocol"]
+            guid_name = mapping["name"]
+
+            if protocol_type not in bv.types:
+                log_warn(f"{protocol_type} not found in binary view. Your version of Binary Ninja may be out of data. Please update to newer versions.")
+
+            protocols[guid] = (protocol_type, guid_name)
+
+    if os.path.exists(user_mapping_path):
+        continue_text = "\nContinue without user-provided protocol bindings?"
+        try:
+            with open(user_mapping_path, "r") as f:
+                user_mapping = json.load(f)
+                if not user_mapping:
+                    raise AssertionError("User protocol mapping file exists but is empty")
+
+                for mapping in user_mapping:
+                    guid = struct.pack("<IHH8B", *mapping["guid"])
+                    protocol_type = mapping["protocol"]
+                    guid_name = mapping["name"]
+                    if protocol_type not in bv.types:
```

Comment:
We should not require that the types are in `bv.types` because we can still use the type name associated with the GUID to name variables.